src/ring_theory/ore_localization/ore_set.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2022 Jakob von Raumer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakob von Raumer, Kevin Klinge
+-/
+import group_theory.subgroup.basic
+
+/-!
+
+# (Right) Ore sets
+
+This defines right Ore sets on arbitrary monoids.
+
+## References
+
+* https://ncatlab.org/nlab/show/Ore+set
+
+-/
+
+namespace ore_localization
+
+section monoid
+
+/-- A submonoid `S` of a monoid `R` is (right) Ore if common factors on the left can be turned
+into common factors on the right, and if each pair of `r : R` and `s : S` admits an Ore numerator
+`v : R` and an Ore denominator `u : S` such that `r * u = s * v`. -/
+class ore_set {R : Type*} [monoid R] (S : submonoid R) :=
+  (ore_left_cancel : ∀ (r₁ r₂ : R) (s : S), ↑s * r₁ = s * r₂ → {s' : S // r₁ * s' = r₂ * s'})
+  (ore_num : R → S → R)
+  (ore_denom : R → S → S)
+  (ore_eq : ∀ (r : R) (s : S), r * (ore_denom r s) = s * (ore_num r s))
+
+variables {R : Type*} [monoid R] {S : submonoid R} [ore_set S]
+
+/-- Common factors on the left can be turned into common factors on the right, a weak form of
+cancellability. -/
+def ore_left_cancel (r₁ r₂ : R) (s : S) (h : ↑s * r₁ = s * r₂) : {s' : S // r₁ * s' = r₂ * s'} :=
+ore_set.ore_left_cancel r₁ r₂ s h
+
+/-- The common right factor of an Ore set's weak cancellability property. -/
+def ore_left_cancel_factor (r₁ r₂ : R) (s : S) (h : (s : R) * r₁ = s * r₂) : S :=
+(ore_left_cancel r₁ r₂ s h).1
+
+/-- The defining equality of an Ore set's weak cancellability. -/
+lemma ore_left_cancel_factor_eq (r₁ r₂ : R) (s : S) (h : (s : R) * r₁ = s * r₂) :
+  r₁ * (ore_left_cancel_factor r₁ r₂ s h) = r₂ * (ore_left_cancel_factor r₁ r₂ s h) :=
+(ore_left_cancel r₁ r₂ s h).2
+
+/-- The Ore numerator of a fraction. -/
+def ore_num (r : R) (s : S) : R := ore_set.ore_num r s
+
+/-- The Ore denominator of a fraction. -/
+def ore_denom (r : R) (s : S) : S := ore_set.ore_denom r s
+
+/-- The Ore condition of a fraction, expressed in terms of `ore_num` and `ore_denom`. -/
+lemma ore_eq (r : R) (s : S) : r * (ore_denom r s) = s * (ore_num r s) := ore_set.ore_eq r s
+
+/-- The Ore condition bundled in a sigma type. This is useful in situations where we want to obtain
+both witnesses and the condition for a given fraction. -/
+def ore_condition (r : R) (s : S) : Σ' r' : R, Σ' s' : S, r * s' = s * r' :=
+⟨ore_num r s, ore_denom r s, ore_eq r s⟩
```

Comment:
I'm not convinced this is useful; can't the caller just use `ore_num r s`, `ore_denom r s`, and `ore_eq r s` independently? I don't think this pattern is one that mathlib uses anywhere else.

Reply:
It's used in the later construction quite a lot. At first I wanted to avoid it but it just saves a lot of `set` and `replace` in the proofs. We always want short identifiers for all the three components.